esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

robert-hh opened the pull request:
```
esp8266/moduos.c:
    added proxy stubs for stat() and chdir(), and an alias for rmdir()
extmod/vfs_fat.c:
    - added the functions fo stat() and chdir() and an alias for rmdir()
    - replaced the text error messages by the Posix error numbers using
      the mechanism already provided
```

robert-hh committed:
esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

    esp8266/moduos.c:
        added proxy stubs for stat() and chdir(), and an alias for rmdir()
    extmod/vfs_fat.c:
        - added the functions fo stat() and chdir() and an alias for rmdir()
        - replaced the text error messages by the Posix error numbers using
          the mechanism already provided

pfalcon commented on esp8266/moduos.c:
```diff
@@ -38,6 +38,7 @@
 #include "esp_mphal.h"
 #include "user_interface.h"
 
+
```
Please avoid whitespace/newline changes like this.

pfalcon commented on extmod/vfs_fat.c:
```diff
@@ -35,6 +35,9 @@
 #include "extmod/fsusermount.h"
 
+#include "timeutils.h"
+
```
These 2 empty lines aren't needed, includes likely should just go together, with an empty line after them.

pfalcon commented on extmod/vfs_fat.c:
```diff
@@ -111,22 +114,123 @@ STATIC mp_obj_t fat_vfs_mkdir(mp_obj_t vfs_in, mp_obj_t path_o) {
+//   path_equal(/, /) -> true
+// second argument must be in canonical form (meaning no trailing slash, unless it's just /)
+STATIC bool path_equal(const char *path, const char *path_canonical) {
+    for (; *path_canonical != '\0' && *path == *path_canonical; ++path, ++path_canonical) {
```
Well, that's interesting pattern of using for (). I guess, even @dpgeorge would agree that instead of empty block, this calls for just comma, i.e.:
~~~
for (; *path_canonical != '\0' && *path == *path_canonical; ++path, ++path_canonical);
~~~

pfalcon replied:
But I'd use while still (though sometimes I used for in such manner too).

robert-hh replied:
That was just copied in from stmhal code

robert-hh replied:
And this functions turns up in the WiPy port too, looking similar.

pfalcon replied:
Well, sounds good then ;-).

robert-hh commented:
Is that just a hint, which I will consider, or shall I update the files?

robert-hh commented:
Hi Paul. Is that just a hint, which I will consider, or shall I update the files?

pfalcon commented:
> Is that just a hint, which I will consider, or shall I update the files?

Generally, I give comments about things which I wouldn't merge with or not sure about. You can agree and fix, or argue why it should be done like you did, or just leave for someone else to review. In this case, the rest of the code looks ok, and it's generally needed features (and in my queue to be fixed too), so having a quick turnaround on this is a godo idea.

robert-hh commented:
I'll change that later today. 
I also found a glitch just now, when doing uos.stat(".") in root, it returns error 22. That's what all ports do. I could fix that, by calling f_getcwd() and checking, whether it's root. That'll take a few minutes longer.

pfalcon commented:
Thanks, and no hurry of course. If all ports do that, I'd leave it for later, and for sure as a separate git commit. And if fixed, it rather would be fixed for all ports together.

robert-hh committed:
esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

    esp8266/moduos.c:
        added proxy stubs for stat() and chdir(), and an alias for rmdir()
    extmod/vfs_fat.c:
        - added the functions fo stat() and chdir() and an alias for rmdir()
        - replaced the text error messages by the Posix error numbers using
          the mechanism already provided
        - coding style fixe

robert-hh commented:
I updated the code style fixes. On check above failed, but the details look unrelated, like a fail in the test system.

robert-hh committed:
esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

    esp8266/moduos.c:
        added proxy stubs for stat() and chdir(), and an alias for rmdir()
    extmod/vfs_fat.c:
        - added the functions fo stat() and chdir() and an alias for rmdir()
        - replaced the text error messages by the Posix error numbers using
          the mechanism already provided
        - changed the style loops.

robert-hh committed:
esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

    esp8266/moduos.c:
        added proxy stubs for stat() and chdir(), and an alias for rmdir()
    extmod/vfs_fat.c:
        - added the functions fo stat() and chdir() and an alias for rmdir()
        - replaced the text error messages by the Posix error numbers using
          the mechanism already provided
        - changed coding style
    unix/Makefile:
        added lib/timeutils.h and lib.timeutils.c to the files set

robert-hh committed:
esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

    esp8266/moduos.c:
        added proxy stubs for stat() and chdir(), and an alias for rmdir()
    extmod/vfs_fat.c:
        - added the functions fo stat() and chdir() and an alias for rmdir()
        - replaced the text error messages by the Posix error numbers using
          the mechanism already provided
        - changed coding style & Travis CI warnings
    unix/Makefile:
        added lib/timeutils.h and lib.timeutils.c to the files set

robert-hh committed:
esp8266: Add stat(), chdir() and rmdir() to module uos and vfs

    esp8266/moduos.c:
        added proxy stubs for stat() and chdir(), and an alias for rmdir()
    extmod/vfs_fat.c:
        - added the functions fo stat() and chdir() and an alias for rmdir()
        - replaced the text error messages by the Posix error numbers using
          the mechanism already provided
        - changed coding style & (hopefully all) Travis CI warnings
    unix/Makefile:
        added lib/timeutils.h and lib.timeutils.c to the files set

robert-hh commented:
The travis CI check constantly fails in the nanbox build. Since the functions which fails is from stmhal, I wonder why it passed before. I have to dig further.

dpgeorge commented on extmod/vfs_fat.c:
```diff
@@ -111,22 +113,127 @@ STATIC mp_obj_t fat_vfs_mkdir(mp_obj_t vfs_in, mp_obj_t path_o) {
+    t->items[8] = MP_OBJ_NEW_SMALL_INT(seconds); // st_mtime
+    t->items[9] = MP_OBJ_NEW_SMALL_INT(seconds); // st_ctime
+
+    return (mp_obj_t)t;
```
@robert-hh you need to write it like this: `return MP_OBJ_FROM_PTR(t);`

robert-hh replied:
Hello Damien, thank you for the hint. By comparing my code with others,
I came along this Macro too. But I closed (to be re-opened) my PL for
two reasons;
a) The whole approach of vfs_fat looks a little bit WIP. I was puzzled
about the unused parameter mp_obj_t vfs_in of every function, I wanted
to understand it's intention. And I'm still nit sure whether the stat()
function should reside in vfs_fat,c or moduos.c.
b) I messed up my copy of the git repository.
Best Regards, Robert


On 27.05.2016 17:48, Damien George wrote:
> In extmod/vfs_fat.c
> <https://github.com/micropython/micropython/pull/2110#discussion_r64925530>:
> 
>> +        (fno.ftime >> 11) & 0x1f,
>> +        (fno.ftime >> 5) & 0x3f,
>> +        2 * (fno.ftime & 0x1f)
>> +    );
>> +    t->items[0] = MP_OBJ_NEW_SMALL_INT(mode); // st_mode
>> +    t->items[1] = MP_OBJ_NEW_SMALL_INT(0); // st_ino
>> +    t->items[2] = MP_OBJ_NEW_SMALL_INT(0); // st_dev
>> +    t->items[3] = MP_OBJ_NEW_SMALL_INT(0); // st_nlink
>> +    t->items[4] = MP_OBJ_NEW_SMALL_INT(0); // st_uid
>> +    t->items[5] = MP_OBJ_NEW_SMALL_INT(0); // st_gid
>> +    t->items[6] = MP_OBJ_NEW_SMALL_INT(fno.fsize); // st_size
>> +    t->items[7] = MP_OBJ_NEW_SMALL_INT(seconds); // st_atime
>> +    t->items[8] = MP_OBJ_NEW_SMALL_INT(seconds); // st_mtime
>> +    t->items[9] = MP_OBJ_NEW_SMALL_INT(seconds); // st_ctime
>> +
>> +    return (mp_obj_t)t;
> 
> @robert-hh <https://github.com/robert-hh> you need to write it like
> this: |return MP_OBJ_FROM_PTR(t);|
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/micropython/micropython/pull/2110/files/172a3fb51f31f4789e385828c57326417ecc8cdc#r64925530>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AL5hxMBXKUUJ_0Zt2ZX0hiUNw-UP7B8wks5qFxIwgaJpZM4Imtsp>.
>

pfalcon replied:
there's also c): please-please, don't try to make several changes at once! "stat(), chdir() and rmdir()", should be (at least) 3 separate pull requests. No need to start with next one until previous is finished, submitted, ran thru CI, reviewed, merged. @robert-hh , your previous PR was the right way to do it. It's better to spend a day on each, than a week (or more) on something which tries to pack them together.

robert-hh replied:
OK. I'll start with preparation - replacing the text error messages in
vfs_fat.c with the Posix number messages.

On 27.05.2016 21:22, Paul Sokolovsky wrote:
> In extmod/vfs_fat.c
> <https://github.com/micropython/micropython/pull/2110#discussion_r64955190>:
> 
>> +        (fno.ftime >> 11) & 0x1f,
>> +        (fno.ftime >> 5) & 0x3f,
>> +        2 * (fno.ftime & 0x1f)
>> +    );
>> +    t->items[0] = MP_OBJ_NEW_SMALL_INT(mode); // st_mode
>> +    t->items[1] = MP_OBJ_NEW_SMALL_INT(0); // st_ino
>> +    t->items[2] = MP_OBJ_NEW_SMALL_INT(0); // st_dev
>> +    t->items[3] = MP_OBJ_NEW_SMALL_INT(0); // st_nlink
>> +    t->items[4] = MP_OBJ_NEW_SMALL_INT(0); // st_uid
>> +    t->items[5] = MP_OBJ_NEW_SMALL_INT(0); // st_gid
>> +    t->items[6] = MP_OBJ_NEW_SMALL_INT(fno.fsize); // st_size
>> +    t->items[7] = MP_OBJ_NEW_SMALL_INT(seconds); // st_atime
>> +    t->items[8] = MP_OBJ_NEW_SMALL_INT(seconds); // st_mtime
>> +    t->items[9] = MP_OBJ_NEW_SMALL_INT(seconds); // st_ctime
>> +
>> +    return (mp_obj_t)t;
> 
> there's also c): please-please, don't try to make several changes at
> once! "stat(), chdir() and rmdir()", should be (at least) 3 separate
> pull requests. No need to start with next one until previous is
> finished, submitted, ran thru CI, reviewed, merged. @robert-hh
> <https://github.com/robert-hh> , your previous PR was the right way to
> do it. It's better to spend a day on each, than a week (or more) on
> something which tries to pack them together.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/micropython/micropython/pull/2110/files/172a3fb51f31f4789e385828c57326417ecc8cdc#r64955190>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AL5hxIaJZcn-n4IfnjAoqUHuKjRdh_bwks5qF0R_gaJpZM4Imtsp>.
>
